### PR TITLE
Fix union of generators having no return type 

### PR DESCRIPTION
**Pull request by omarsilva1**
Fixes https://github.com/python/mypy/issues/15141

This adds a check if `iter_type` is a union type so that `expr_type` is set to the generator return type instead of none.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
(the timing change is probably noise)

File: test-data/unit/check-statements.test [hunk 1]
```diff
@@ -2195,6 +2195,14 @@ class B: pass
 def foo(x: int) -> Union[Generator[A, None, None], Generator[B, None, None]]:
     yield x  # E: Incompatible types in "yield" (actual type "int", expected type "Union[A, B]")
 
+[case testReturnTypeOfGeneratorUnion]
+from typing import Generator, Union
+
+class T: pass
+
+def foo(arg: Union[Generator[int, None, T], Generator[str, None, T]]) -> Generator[Union[int, str], None, T]:
```
Comment (JelleZijlstra) [inline test-data/unit/check-statements.test:2203]: What about a union of non-generator types?

Comment (omarsilva1) [reply test-data/unit/check-statements.test:2203]: Yield from a union of non-generator types returns an error (yield from can't be applied to) on line 5100, I added a test for that now.

File: test-data/unit/check-statements.test [hunk 2]
```diff
@@ -2195,6 +2195,23 @@ class B: pass
...
+    return (yield from arg)
+
+[case testYieldFromError]
+from typing import Generator, Union
+
+class T: pass
+class R: pass
+
+def foo(arg: Union[T, R]) -> Generator[Union[int, str], None, T]:
```
Comment (JelleZijlstra) [inline test-data/unit/check-statements.test:2212]: Thanks! What about a Union involving a Generator and some other sort of iterable?

Comment (omarsilva1) [reply test-data/unit/check-statements.test:2212]: I added a test for the union of a generator and iterable

File: mypy/checkexpr.py [hunk 3]
```diff
@@ -5095,7 +5095,11 @@ def visit_yield_from_expr(self, e: YieldFromExpr, allow_none_return: bool = Fals
 
         # Determine the type of the entire yield from expression.
         iter_type = get_proper_type(iter_type)
-        if isinstance(iter_type, Instance) and iter_type.type.fullname == "typing.Generator":
+        if (
+            isinstance(iter_type, Instance)
+            and iter_type.type.fullname == "typing.Generator"
+            or isinstance(iter_type, UnionType)
```
Comment (ikonst) [inline mypy/checkexpr.py:5101]: 10,000ft / 30,000m high observation:

This reads unbalanced to me: one case (`typing.Generator`) is very specific to generators, while the other case is broad (a union of anything).

Maybe the condition is logic that should move into `get_generator_return_type`?

**Comment by hauntsaninja** [comment]
Thanks for working on this! I think ikonst's observation is a good one, e.g. what if we have a Union of Iterator and Generator or something.

I was playing around with this diff:
<details>

```diff
diff --git a/mypy/checker.py b/mypy/checker.py
index c1c31538b..6f1705cf7 100644
--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -952,8 +952,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             # AwaitableGenerator, Generator: tr is args[2].
             return return_type.args[2]
         else:
-            # Supertype of Generator (Iterator, Iterable, object): tr is any.
-            return AnyType(TypeOfAny.special_form)
+            # We have a supertype of Generator (Iterator, Iterable, object)
+            # Treat `Iterator[X]` as a shorthand for `Generator[X, Any, None]`.
+            return NoneType()
 
     def visit_func_def(self, defn: FuncDef) -> None:
         if not self.recurse_into_functions:
diff --git a/mypy/checkexpr.py b/mypy/checkexpr.py
index cd0ff1100..34dcbdef0 100644
--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5177,17 +5177,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
         # Determine the type of the entire yield from expression.
         iter_type = get_proper_type(iter_type)
-        if isinstance(iter_type, Instance) and iter_type.type.fullname == "typing.Generator":
-            expr_type = self.chk.get_generator_return_type(iter_type, False)
-        else:
-            # Non-Generators don't return anything from `yield from` expressions.
-            # However special-case Any (which might be produced by an error).
-            actual_item_type = get_proper_type(actual_item_type)
-            if isinstance(actual_item_type, AnyType):
-                expr_type = AnyType(TypeOfAny.from_another_any, source_any=actual_item_type)
-            else:
-                # Treat `Iterator[X]` as a shorthand for `Generator[X, None, Any]`.
-                expr_type = NoneType()
+        expr_type = self.chk.get_generator_return_type(iter_type, False)
 
         if not allow_none_return and isinstance(get_proper_type(expr_type), NoneType):
             self.chk.msg.does_not_return_value(None, e)
```
</details>

It causes two other tests to fail, but I think the new behaviour is better than the old behaviour. What do you think?

File: mypy/checkexpr.py (hunk 3 above)
Comment (omarsilva1) [reply mypy/checkexpr.py]: Thanks for the observation, I've committed something that uses the suggestion from @hauntsaninja and fixed the failing test.

**Comment by omarsilva1** [comment]
Thanks for the feedback! I added your suggestions, I also think that this new behaviour is more natural than the previous one. 

I only had one test failing locally, where `Iterator[int]` was the return type of a function. Should be fixed now.

**Comment by hauntsaninja** [comment]
Any luck with fixing the test?

**Pull request closed without merging**